compiler/src/yul/mappers/expressions.rs

```diff
@@ -0,0 +1,278 @@
+use crate::errors::CompileError;
+use crate::yul::namespace::scopes::{ContractDef, FunctionDef, FunctionScope, Shared};
+use crate::yul::namespace::types::{Base, FixedSize};
+use std::rc::Rc;
+use vyper_parser::ast as vyp;
+use vyper_parser::span::Spanned;
+use yultsur::*;
+
+/// Builds a Yul expression from a Vyper expression.
+pub fn expr(
+    scope: Shared<FunctionScope>,
+    expr: &vyp::Expr,
+) -> Result<(yul::Expression, FixedSize), CompileError> {
+    match expr {
+        vyp::Expr::Name(name) => expr_name(scope, name.to_string()),
+        vyp::Expr::Num(num) => Ok((literal_expression! {(num)}, FixedSize::Base(Base::U256))),
+        vyp::Expr::Subscript { value, slices } => expr_subscript(scope, &value.node, &slices.node),
+        vyp::Expr::Attribute { value, attr } => {
+            expr_attribute(scope, &value.node, attr.node.to_string())
+        }
+        _ => unimplemented!("Expression"),
+    }
+}
+
+/// Retrieves the &str value of a name expression.
+pub fn expr_name_str<'a>(expr: &'a vyp::Expr<'a>) -> Result<&'a str, CompileError> {
+    if let vyp::Expr::Name(name) = expr {
+        return Ok(*name);
+    }
+
+    Err(CompileError::static_str("Not a name expression."))
+}
+
+/// Retrieves the &str value of a name expression and converts it to a String.
+pub fn expr_name_string(expr: &vyp::Expr) -> Result<String, CompileError> {
+    expr_name_str(expr).map(|name| name.to_string())
+}
+
+/// Builds a Yul expression from the first slice, if it is an index.
+pub fn slices_index(
+    scope: Shared<FunctionScope>,
+    slices: &Vec<Spanned<vyp::Slice>>,
+) -> Result<(yul::Expression, FixedSize), CompileError> {
+    if let Some(first_slice) = slices.first() {
+        if let vyp::Slice::Index(index) = &first_slice.node {
+            return Ok(expr(scope, index)?);
+        }
+
+        return Err(CompileError::static_str("First slice is not an index"));
+    }
+
+    return Err(CompileError::static_str("No slices in vector"));
+}
+
+fn expr_name(
+    scope: Shared<FunctionScope>,
+    name: String,
+) -> Result<(yul::Expression, FixedSize), CompileError> {
+    match scope.borrow().def(name.clone()) {
+        Some(FunctionDef::Base(base)) => {
+            Ok((identifier_expression! {(name)}, FixedSize::Base(base)))
+        }
+        Some(FunctionDef::Array(array)) => {
+            Ok((identifier_expression! {(name)}, FixedSize::Array(array)))
+        }
+        None => Err(CompileError::static_str("No definition found")),
+    }
+}
+
+// FIXME: The left side should be evaluated and then indexed.
```

Comment:
Let me expand on this a bit.

Say we have the expression `self.messages[5]`, where `self.messages` is a `map<u256, bytes[100]>` declared in the contract scope. Currently, the `expr_subscript` mapper function takes the entire expression and filters it through a series of nested `expr_subscript_` calls that resolve to a `(yul::Expression, FixedSize)` tuple. It works for guest_book.vy, but is extremely limited in that you have to hard-code every possible left side (`self.<contract def>` in this case) to implement the language's desired behavior. I believe I just copied this pattern from the assignments module where it works better.

The `FIXME` comment vaguely implies that we should instead use the normal `expr` mapper function to map the left side and then combine it with the separately mapped slices to form a full expression, instead of hard-coding a bunch of nested functions. In order to do this though, we need to extend the `(yul::Expression, FixedSize)` tuple to include some more information about type and location. Possibly adding a struct like the one bellow and making it the default return for each expression mapper.

```rust
/// Extended expression
pub struct ExtExpression {
    pub expression: yul::Expression,
    pub location: Location,
    pub typ: Type,
}
```

In the case of `self.messages`, we would have an extended expression that looks something like this:

```
expression: 0 // storage index
location: Storage
type: Map<u256, bytes[100]>
```

With this information, we would be able to form the final extended expression for `self.messages[5]` in `expr_subscript` by combining the left (`self.messages`) and right (`[5]`) side into:

```
expression: dualkeccak(0, 5) // indexed key
location: Storage
type: bytes[100]
```

This way we can actually fully implement the language and remove unnecessary code in the expressions module.

Hopefully this clears things up, I was pretty confused by the comment at first.